6-infrastructure-configuration-management/6.1.4-terraform-providers.md

```diff
@@ -14,49 +14,199 @@ Throughout this section it will be helpful to refer to the [Custom Framework Pro
+   - A couple things to keep in mind here: 
+      1. The tutorial has a lot of extra checking and setting of variables you might not necessarily need
+      2. Their struct that holds the data for the provider configuration includes things for authentication which the API we are using does not have
+      3. Instead of having their client be an "http.DefaultClient" like in the boilerplate, they have a wrapper around that to keep all the other data they need for it close by and is probably something you should look into replicating.  A link to their implimentation is [here](https://github.com/hashicorp-demoapp/hashicups-client-go)
+
+### Challenge:
+- Ensure you write tests for Exercise 4 and 5 that go in depth to how the resource and data source should work before running terraform plan and apply.
+
+!> Do not think about this last minute! Recommended to think about testing early. Helps with planning and designing how the provider will be built before beginning to build the rest of the resources and datasources.
+
+### Exercise 4: Implementing Engineer resource and datasource
+
+?> Some basic testing will be provided for engineers as a starting point for the challenge. It is encouraged to include more testing for your resources and datasources to ensure proper functionality for what was envisioned.
+
+   1. Create a model for the engineers resource using the Terraform types.
+
+   1. Think about what the Schema should look like:  
+    - What does the data the API returns going to look like?
+    - What variables does the API need vs what variables will be given to us?
+
+   1. Impliment all of the required CRUD operations using the tutorial as a reference.  
+    - In the documentation they use methods within their client wrapper to perform the API calls:
+        - This is time where you should plan how you want to organize these for the rest of your resources.
+        - Recommendation: Making a wrapper as done in the tutorial is recommended to make the code easier to read and understand.
+
+?> For this section ignore how the datasource is implimented in the tutorial as it is an incredibly uncommon approach, your datasource should be 1:1 to another resource of the same type.  Datasources are used to get information about resources not managed by terraform, and given that they arent being managed, you will lack certian information about them such as an ID so you will need to look at the API for another way to grab the desired resources. 
+
+?> Also think about what should happen when Terraform goes to fetch information about a resource it has inside its state but cant find in the API.  What does that mean and how should it be corrected?
+
+### Exercise 5: Implementing Dev, Ops, DevOps resource and datasource
+
+Implementing the next resoruces are going to be similar to the implementation of the engineer resoruce. Refer back to the section [API Overview](#api-overview)
+
+1. Begin with implementing the read functionality
```

Comment:
Is there more steps intended to be added here?

Reply:
Exercise 5 for the Dev, Ops and DevOps resource and data sources will have more steps. This is more of a repeat of engineer so we're wanting to ensure Exercise 4 (the engineer resource and data source) has more details explaining how to make the resource and test it.